trame_server/state.py

```diff
@@ -274,52 +302,53 @@ def flush(self):
         previous value or if `dirty` has been flagged on the variable and it has
         not been unflagged since.
         """
-        if not self.is_ready:
+        if not self.is_ready or self.is_flushing:
             return None
 
         keys = set()
-        if len(self._pending_update):
-            _keys = set(self._pending_update.keys())
-
-            while len(_keys):
-                keys |= _keys
-
-                # update modified keys for current update batch
-                self._modified_keys.clear()
-                self._modified_keys |= _keys
-
-                # Do the flush
-                if self._push_state_fn:
-                    self._push_state_fn(self._pending_update)
-                self._pushed_state.update(self._pending_update)
-                self._pending_update.clear()
-
-                # Execute state listeners
-                self._state_listeners.add_all(_keys)
-                for fn, translator in self._state_listeners:
-                    if isinstance(fn, weakref.WeakMethod):
-                        callback = fn()
-                        if callback is None:
-                            continue
-                    else:
-                        callback = fn
-
-                    if self._hot_reload:
-                        if not inspect.iscoroutinefunction(callback):
-                            callback = reload(callback)
-
-                    reverse_translated_state = translator.reverse_translate_dict(
-                        self._pushed_state
-                    )
-                    coroutine = callback(**reverse_translated_state)
-                    if inspect.isawaitable(coroutine):
-                        asynchronous.create_task(coroutine)
-
-                self._state_listeners.clear()
-
-                # Check if state change from state listeners
+        with self.flushing():
+            if len(self._pending_update):
                 _keys = set(self._pending_update.keys())
 
+                while len(_keys):
+                    keys |= _keys
```

Comment:
Indentation is starting to be a bit hard to read (IMO).
It would probably be a bit cleaner to have another method for this scope.

```python
    while self.has_pending_update():
        keys |= self._flush_pending_keys()
    ...
    
def has_pending_update(self) -> bool:
    return bool(self._pending_update)
    
def _flush_pending_keys(self) -> set[str]:
    _keys = set(self._pending_update.keys())
    ...
    return _keys   

```

First `if len(self._pending_update)` can be removed with this change as well